native/src/lib.rs

```diff
@@ -81,11 +85,107 @@ declare_types! {
                                 None)
                 .expect("Query failed");
 
-            Ok(try!(JsString::new_or_throw(scope, &results.len().to_string()[..])).upcast())
+            if let &QueryResults::Scalar(Some(TypedValue::Keyword(ref rc))) = results {
+                println!("TODO: Matched Scalar: {:?}", rc);
+            }
+
+            if let &QueryResults::Rel(ref rel) = results {
+                let iter = rel.iter();
+                let array: Handle<JsArray> = JsArray::new(scope, iter.len() as u32);
+                let mut i = 0;
+                for r in rel {
+                    let r_iter = r.iter();
+                    let r_array: Handle<JsArray> = JsArray::new(scope, r_iter.len() as u32);
+                    let mut j = 0;
+                    for item in r_iter {
+                        let processed_value = process_typed_value(scope, item);
+                        match processed_value {
+                            ReturnedHandle::JsString(handle) => {
+                                try!(r_array.set(j, handle));
+                            }
+                            ReturnedHandle::JsBoolean(handle) => {
+                                try!(r_array.set(j, handle));
+                            }
+                            ReturnedHandle::JsNumber(handle) => {
+                                try!(r_array.set(j, handle));
+                            }
+                        }
+
+                        j = j + 1;
+                    }
+                    try!(array.set(i, r_array));
+                    i = i + 1;
+                    try!(output.set("results", array));
+                }
+            }
+
+            if let &QueryResults::Tuple(Some(ref tuple)) = results {
+                println!("TODO: Matched Tuple: {:?}", tuple);
+            }
+
+            if let &QueryResults::Coll(ref coll) = results {
+                // println!("Matched Coll: {:?}", coll);
+                let iter = coll.iter();
+                let array: Handle<JsArray> = JsArray::new(scope, iter.len() as u32);
+                let mut i = 0;
+                for item in iter {
+                    let processed_value = process_typed_value(scope, item);
+
+                    // TODO: Can we match all possible types with a single pattern
+                    match processed_value {
+                        ReturnedHandle::JsString(handle) => {
+                            try!(array.set(i, handle));
+                        }
+                        ReturnedHandle::JsBoolean(handle) => {
+                            try!(array.set(i, handle));
+                        }
+                        ReturnedHandle::JsNumber(handle) => {
+                            try!(array.set(i, handle));
+                        }
+                    }
+                    i = i+1;
```

Comment:
Rather than manually indexing:

```
let mut i = 0;
for x in items.iter() {
    …
    i += 1;
}
```

use `enumerate`:

```
for (i, x) in items.iter().enumerate() {
    …
}
```